File: py/src/precompiled/treenode.py
from dataclasses import dataclass
from typing import Optional, List

from .treenode_base import array_to_tree


@dataclass
class TreeNode(object):
    val: int
    left: Optional['TreeNode']
    right: Optional['TreeNode']

    def __init__(self, val, left: Optional['TreeNode'] = None, right: Optional['TreeNode'] = None):
        self.val = val
        self.left = left
        self.right = right

    def set_left(self, n: 'TreeNode'):
        self.left = n

    def set_right(self, n: 'TreeNode'):
        self.right = n

    @staticmethod
    def array_to_tree(arr: List[Optional[int]]) -> Optional['TreeNode']:
        return array_to_tree(TreeNode, arr)

File: cpp/precompiled/TreeNode.cpp
#include "TreeNode.h"

// Convert array to binary tree
TreeNode *TreeNode::arrayToTree(const std::vector<std::optional<int>> &vals){
  if (vals.empty()){
    return nullptr;
  }

  TreeNode *root = new TreeNode(vals[0].value());
  std::deque<TreeNode *> queue;
  queue.push_back(root);
  size_t i = 1;
  size_t length = vals.size();

  while (!queue.empty() && i < length){
    TreeNode *node = queue.front();
    queue.pop_front();

    if (i < length && vals[i]){
      TreeNode *left = new TreeNode(vals[i].value());
      node->left = left;
      queue.push_back(left);
    }
    i++;

    if (i < length && vals[i]){
      TreeNode *right = new TreeNode(vals[i].value());
      node->right = right;
      queue.push_back(right);
    }
    i++;
  }

  return root;
}